File: tutorials/tutorial_06_video_event/python/main.py
########################################################################
#
# Copyright (c) 2023, STEREOLABS.
#
# All rights reserved.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
########################################################################

import pyzed.sl as sl
import pyzed.sl_hub as hub


def main():
    # Create camera object
    zed = sl.Camera()

    # Initialize the communication to ZED Hub, with a zed camera.
    status_hub = hub.HubClient.connect("event_app")
    if status_hub != hub.STATUS_CODE.SUCCESS:
        print("Initialization error ", status_hub)
        exit(1)

    # Open the zed camera
    init_params = sl.InitParameters()
    init_params.camera_resolution = sl.RESOLUTION.HD2K
    init_params.depth_mode = sl.DEPTH_MODE.PERFORMANCE
    status_zed = zed.open(init_params)
    if status_zed != sl.ERROR_CODE.SUCCESS:
        hub.HubClient.send_log(
            "Camera initialization error : " + str(status_zed), hub.LOG_LEVEL.ERROR)
        exit(1)

    # Register the camera once it's open
    update_params = hub.UpdateParameters()
    status_hub = hub.HubClient.register_camera(zed, update_params)
    if status_hub != hub.STATUS_CODE.SUCCESS:
        print("Camera registration error ", status_hub)
        exit(1)

    # Enable Position tracking (mandatory for object detection)
    track_params = sl.PositionalTrackingParameters()
    track_params.set_as_static = False
    status_zed = zed.enable_positional_tracking(track_params)
    if status_zed != sl.ERROR_CODE.SUCCESS:
        hub.HubClient.send_log(
            "Positional tracking initialization error : " + str(status_zed), hub.LOG_LEVEL.ERROR)
        exit(1)

    # Enable the Objects detection module
    object_detection_params = sl.ObjectDetectionParameters()
    object_detection_params.image_sync = True
    object_detection_params.enable_tracking = False
    object_detection_params.detection_model = sl.OBJECT_DETECTION_MODEL.MULTI_CLASS_BOX_FAST
    status_zed = zed.enable_object_detection(object_detection_params)
    if status_zed != sl.ERROR_CODE.SUCCESS:
        hub.HubClient.send_log(
            "Object detection initialization error : " + str(status_zed), hub.LOG_LEVEL.ERROR)
        exit(1)

    # Object Detection runtime parameters : detect person only
    # see  the ZED Doc for the other available classes https://www.stereolabs.com/docs/api/group__Object__group.html#ga13b0c230bc8fee5bbaaaa57a45fa1177 
    object_detection_runtime_params = sl.ObjectDetectionRuntimeParameters()
    object_detection_runtime_params.detection_confidence_threshold = 50
    object_detection_runtime_params.object_class_filter = []
    object_detection_runtime_params.object_class_filter.append(
        sl.OBJECT_CLASS.PERSON)

    counter_no_detection = 0
    objects = sl.Objects()
    event_reference = ""
    first_event_sent = False
    last_ts = sl.Timestamp()

    # Main loop
    while True:
        status_zed = zed.grab()
        if status_zed != sl.ERROR_CODE.SUCCESS:
            break

        zed.retrieve_objects(objects, object_detection_runtime_params)

        #----------     Define event   ----------
        # Let's define a video event as a video on which you can detect someone at least every 10 frames.
        # If nobody is detected for 10 frames, a new event is defined next time someone is detected.
        # Cf README.md to understand how to use the event_reference to define a new event.

        current_ts = objects.timestamp
        if len(objects.object_list) >= 1 and (current_ts.get_seconds() - last_ts.get_seconds() >= 2):
            new_event = True
            if counter_no_detection >= 10 or not first_event_sent:
                event_reference = "detected_person_" + \
                    str(current_ts.get_milliseconds())
            else:
                new_event = False

            event_params = hub.EventParameters()
            event_params.timestamp = current_ts.get_milliseconds()
            event_params.reference = event_reference
            event_label = "People detection"
            event_to_send = {}
            event_to_send["message"] = "Current event as reference : " + \
                str(event_reference)
            event_to_send["nb_detected_person"] = len(objects.object_list)

            if new_event or not first_event_sent:
                hub.HubClient.start_video_event(
                    zed, event_label, event_to_send)
                first_event_sent = True
                print("Event started")
            else:
                hub.HubClient.update_video_event(
                    zed, event_label, event_to_send)
                print("Event updated")

            counter_no_detection = 0
            last_ts = current_ts

        elif current_ts.get_seconds() - last_ts.get_seconds() >= 2:
            counter_no_detection = counter_no_detection + 1

        # In the end of a grab(), always call a update() on the cloud.
        hub.HubClient.update(zed)

    # Handling camera error
    if status_zed != sl.ERROR_CODE.SUCCESS:
        hub.HubClient.send_log("Grab failed, restarting camera. " + str(status_zed),
                                hub.LOG_LEVEL.ERROR)
        zed.close()
        sl.Camera.reboot(zed.get_camera_information().serial_number)
        
    # Close the camera
    if zed.is_opened():
        zed.close()

    # Close the communication with ZED Hub properly.
    status_hub = hub.HubClient.disconnect()
    if status_hub != hub.STATUS_CODE.SUCCESS:
        print("Terminating error ", status_hub)
        exit(1)

    return


if __name__ == "__main__":
    main()

File: tutorials/tutorial_06_video_event/cpp/main.cpp
#include <stdio.h>
#include <string.h>
#include <chrono>

#include <sl/Camera.hpp>
#include <sl_hub/HubClient.hpp>

using namespace std;
using namespace sl;
using namespace sl_hub;
using json = sl_hub::json;

int main(int argc, char **argv)
{
    // Create camera object
    std::shared_ptr<sl::Camera> p_zed;
    p_zed.reset(new sl::Camera());

    // Initialize the communication to ZED Hub, with a zed camera.
    STATUS_CODE status_hub = HubClient::connect("event_app");
    if (status_hub != STATUS_CODE::SUCCESS)
    {
        std::cout << "HubClient " << status_hub << std::endl;
        exit(EXIT_FAILURE);
    }

    // Open the ZED camera
    sl::InitParameters initParameters;
    initParameters.camera_resolution = RESOLUTION::HD2K;
    initParameters.depth_mode = DEPTH_MODE::PERFORMANCE;
    sl::ERROR_CODE status_zed = p_zed->open(initParameters);
    if (status_zed != ERROR_CODE::SUCCESS)
    {
        HubClient::sendLog("Camera initialization error : " + std::string(toString(status_zed)), LOG_LEVEL::ERROR);
        exit(EXIT_FAILURE);
    }

    // Register the camera once it's open
    UpdateParameters updateParameters;
    status_hub = HubClient::registerCamera(p_zed, updateParameters);
    if (status_hub != STATUS_CODE::SUCCESS) {
        std::cout << "Camera registration error " << status_hub << std::endl;
        exit(EXIT_FAILURE);
    }

    // Enable Position tracking (mandatory for object detection)
    sl::PositionalTrackingParameters track_params;
    track_params.set_as_static = false;
    auto zed_error = p_zed->enablePositionalTracking(track_params);
    if (zed_error != ERROR_CODE::SUCCESS)
    {
        std::cout << sl::toVerbose(zed_error) << "\nExit program." << std::endl;
        p_zed->close();
        return 1;
    }

    // Enable the Objects detection module
    std::cout << "Enable Object Detection Module" << std::endl;
    sl::ObjectDetectionParameters obj_det_params;
    obj_det_params.image_sync = true;
    obj_det_params.enable_tracking = false;
    obj_det_params.detection_model = sl::OBJECT_DETECTION_MODEL::MULTI_CLASS_BOX_FAST;
    zed_error = p_zed->enableObjectDetection(obj_det_params);
    if (zed_error != ERROR_CODE::SUCCESS)
    {
        std::cout << sl::toVerbose(zed_error) << "\nExit program." << std::endl;
        p_zed->close();
        return 1;
    }

    // Object Detection runtime parameters : detect person only
    ObjectDetectionRuntimeParameters objectTracker_parameters_rt;
    objectTracker_parameters_rt.detection_confidence_threshold = 50;
    objectTracker_parameters_rt.object_class_filter.clear();
    objectTracker_parameters_rt.object_class_filter.push_back(sl::OBJECT_CLASS::PERSON);

    // Main loop
    int counter_no_detection = 0;
    sl::Objects objects;
    std::string event_reference = "";
    bool first_event_sent = false;
    sl::Timestamp last_ts = 0;
    while (true)
    {
        // Grab a new frame from the ZED
        status_zed = p_zed->grab();
        if (status_zed != ERROR_CODE::SUCCESS)
            break;

        p_zed->retrieveObjects(objects, objectTracker_parameters_rt);

        /*******     Define event   *********/
        /*
        Let's define a video event as a video on which you can detect someone at least every 10 frames.
        If nobody is detected for 10 frames, a new event is defined next time someone is detected.
        Cf README.md to understand how to use the event_reference to define a new event.
        */

        sl::Timestamp current_ts = objects.timestamp;
        if (objects.object_list.size() >= 1 && current_ts.getSeconds() - last_ts.getSeconds() >= 2)
        {
            bool new_event = true;
            if (counter_no_detection >= 10 || !first_event_sent)
            {
                event_reference = "detected_person_" + std::to_string(current_ts.getMilliseconds());
            }
            else
            {
                // keep previous event reference --> The current frame will be defined as being part of the previous video event
                new_event = false;
            }
            EventParameters event_params;
            event_params.timestamp = current_ts.getMilliseconds();
            event_params.reference = event_reference;
            std::string event_label = "People Detection"; // or label of your choice
            json event2send;                              // Use to store all the data associated to the video event.
            event2send["message"] = "Current event as reference " + event_reference;
            event2send["nb_detected_person"] = objects.object_list.size();

            if (new_event || !first_event_sent)
            {
                HubClient::startVideoEvent(p_zed, event_label, event2send, event_params);
                first_event_sent = true;
                std::cout << "Event started" << std::endl;
            }
            else
            {
                HubClient::updateVideoEvent(p_zed, event_label, event2send, event_params);
                std::cout << "Event updated" << std::endl;
            }
            counter_no_detection = 0; // reset counter as someone as been detected
            last_ts = current_ts;
        }
        else if (current_ts.getSeconds() - last_ts.getSeconds() >= 2)
        {
            counter_no_detection++;
        }
        /*******************************/

        // Always update Hub at the end of the grab loop
        HubClient::update(p_zed);
    }

    // Handling camera error
    if (status_zed != ERROR_CODE::SUCCESS)
    {
        HubClient::sendLog("Grab failed, restarting camera. " + std::string(toString(status_zed)), LOG_LEVEL::ERROR);
        p_zed->close();
        sl::ERROR_CODE e = sl::Camera::reboot(p_zed->getCameraInformation().serial_number);
    }

    // Close the camera
    if (p_zed->isOpened())
        p_zed->close();

    status_hub = HubClient::disconnect();
    if (status_hub != STATUS_CODE::SUCCESS)
    {
        std::cout << "Terminating error " << status_hub << std::endl;
        exit(EXIT_FAILURE);
    }

    return 0;
}
